Entities/MOSRotating.cpp

```diff
@@ -1600,167 +1495,133 @@ bool MOSRotating::DrawMOIDIfOverlapping(MovableObject *pOverlapMO)
-/// <param name="isHardcodedAttachable">Whether or not the Attachable should be treated as hardcoded.</param>
-void MOSRotating::AddAttachable(Attachable *pAttachable, const Vector & parentOffsetToSet, bool isHardcodedAttachable)
-{
-	if (pAttachable)
-	{
-		pAttachable->Attach(this, parentOffsetToSet);
 
-		// Set the attachable's subgroup ID to it's Unique ID to avoid any possible conflicts when adding atoms to parent group.
-		pAttachable->SetAtomSubgroupID(pAttachable->GetUniqueID());
+bool MOSRotating::RemoveAttachable(Attachable *attachable, bool addToMovableMan, bool addBreakWounds) {
+    if (!attachable || !attachable->IsAttached()) {
+        return false;
+    }
+    RTEAssert(attachable->IsAttachedTo(this), "Tried to remove attachable that was attached to another parent (" + (attachable->GetParent() ? attachable->GetParent()->GetModuleAndPresetName() : "ERROR") + ") Attachable (" + attachable->GetModuleAndPresetName() + "), this should never happen!");
 
-		if (pAttachable->CanCollideWithTerrainWhenAttached())
-		{
-			pAttachable->EnableTerrainCollisions(true);
-		}
+    if (m_Attachables.size() > 0) { m_Attachables.remove(attachable); }
+    attachable->SetParent(nullptr);
+    m_AttachableAndWoundMass -= attachable->GetMass();
 
-		if (!isHardcodedAttachable)
-		{
-			m_Attachables.push_back(pAttachable);
-		}
-		m_AllAttachables.push_back(pAttachable);
-	}
+    std::unordered_map<unsigned long, std::function<void(MOSRotating *, Attachable *)>>::iterator hardcodedAttachableMapEntry = m_HardcodedAttachableUniqueIDsAndSetters.find(attachable->GetUniqueID());
+    if (hardcodedAttachableMapEntry != m_HardcodedAttachableUniqueIDsAndSetters.end()) {
+        hardcodedAttachableMapEntry->second(this, nullptr);
+        m_HardcodedAttachableUniqueIDsAndSetters.erase(hardcodedAttachableMapEntry);
+    }
+    
+    if (addBreakWounds) {
+        if (!m_ToDelete && attachable->GetParentBreakWound()) {
+            AEmitter *parentBreakWound = dynamic_cast<AEmitter *>(attachable->GetParentBreakWound()->Clone());
+            if (parentBreakWound) {
+                parentBreakWound->SetEmitAngle((attachable->GetParentOffset() * m_Rotation).GetAbsRadAngle());
+                AddWound(parentBreakWound, attachable->GetParentOffset(), false);
+                parentBreakWound = nullptr;
+            }
+        }
+        if (!attachable->IsSetToDelete() && attachable->GetBreakWound()) {
+            AEmitter *childBreakWound = dynamic_cast<AEmitter *>(attachable->GetBreakWound()->Clone());
+            if (childBreakWound) {
+                childBreakWound->SetEmitAngle(attachable->GetJointOffset().GetAbsRadAngle());
+                attachable->AddWound(childBreakWound, attachable->GetJointOffset());
+                childBreakWound = nullptr;
+            }
+        }
+    }
+    if (!attachable->IsSetToDelete() && attachable->GetDeleteWhenRemovedFromParent()) {
```

Comment:
If I'm remembering right, the explanation for this is kinda complicated and a little wonky. In short, if the Attachable is set to delete when removed from parent, I delete it right away because it saves the extra work of adding it to MovableMan only to delete it on the next frame.

Why isn't IsSetToDelete treated the same way? Well, there are some places in the code where an Attachable is set to delete, but it's still expected to be valid afterwards, cause the normal behaviour here is/was to put it in MovableMan and let that delete it. I may be misremembering, but I think AHuman weapon switching might be one example of this, and I think some stuff with gibbing the parent might be another. Either way, without a bunch more work I couldn't safely delete it right away if it was set to delete, so I added this odd set of conditions.
The simpler option would, of course, be to make it so if it's set to delete when removed from parent it just gets set to delete and is then added to MovableMan like normal. However, as mentioned at the start, that means the potentially quite hefty task of adding it to MovableMan, and I decided to add this small complexity to make it more efficient.

Reply:
Ok, now I understand, thanks for the explanation.
Originally I meant that they should probably both be deleted immediately, but it also seems to me like letting both of them be deleted by MovableMan should not be too expensive. But even if i'm correct then that's probably something to be done in a different pr.

Reply:
The previous version of this comment is what ended up in my email haha
I think I'll put a TODO here mentioning that this should be changed so they both delete immediately, once it's safely doable.